lua/nvim-treesitter/install.lua

```diff
@@ -14,44 +14,80 @@ local repositories = {
   }
 }
 
--- TODO(kyazdani): there might be a better way to get the plugin path
-for _, path in pairs(api.nvim_list_runtime_paths()) do
-  if string.match(path, '.*/nvim%-treesitter') then
-    package_path = path
-    break
+local function get_package_path()
+  for _, path in pairs(api.nvim_list_runtime_paths()) do
+    if string.match(path, '.*/nvim%-treesitter') then
+      return path
+    end
   end
+
+  return nil
+end
+
+local function get_cache_dir()
+  local home = fn.get(fn.environ(), 'HOME')
+  local xdg_cache = fn.get(fn.environ(), 'XDG_CACHE_HOME')
+
+  if xdg_cache == 0 then
+    xdg_cache = home .. '/.cache'
+  end
+
+  if luv.fs_access(xdg_cache, 'RW') then
+    return xdg_cache
+  elseif luv.fs_access('/tmp', 'RW') then
+    return '/tmp'
+  end
+
+  return nil
 end
 
 local function create_compile_command(ft, path, files)
-  -- TODO(kyazdani): should download the parser in $XDG_CACHE_HOME instead of /tmp
-  return "cd /tmp/tree-sitter-"..ft..[[ && \
-    gcc -o parser.so -shared ]]..files..[[ -Os -I./src && \
-    mv parser.so ]]..path.."/parsers/"..ft..[[.so && \
-    rm -rf /tmp/tree-sitter-]]..ft
+  local cache_folder = get_cache_dir()
+  if not cache_folder then
+    return 'Invalid cache rights, $XDG_CACHE_HOME or /tmp should be read/write', nil
+  end
+
+  local project_repo = cache_folder..'/tree-sitter-'..ft
+  local compile_command = "gcc -o parser.so -shared "..files.." -Os -I./src"
```

Comment:
Maybe using `cc` here instead of `gcc` would allow to use both `gcc` and `clang` to compile the parser, and would avoir some errors

Reply:
For my experience maybe you should add `-lstdc++` in the compile command since some tree sitter requires c++ dependency. You can test it out with lua tree sitter, it'll break if `-lstdc++` is not specified.

Reply:
thanks for the tip i didn't know why some parsers didn't want to compile